internal/configs/test_provider_body.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+// Copyright (c) 2023 HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package configs
+
+import (
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+)
+
+var _ hcl.Body = &testProviderBody{}
+
+// testProviderBody is a wrapper around hcl.Body that allows us to use prepared EvalContext for provider configuration.
+type testProviderBody struct {
+	originalBody hcl.Body
+	evalCtx      *hcl.EvalContext
+}
+
+func (c testProviderBody) Content(schema *hcl.BodySchema) (*hcl.BodyContent, hcl.Diagnostics) {
+	content, diags := c.originalBody.Content(schema)
+	if diags.HasErrors() {
+		return nil, diags
+	}
+	attrs := content.Attributes
+	for name, attr := range attrs {
+		vars := attr.Expr.Variables()
+		containsRunRef := false
```

Comment:
Given what we are trying to do here, I think we should try to evaluate and replace all attribute expressions with computed values.  All the variables and run data should be available at this point.  This would also simplify some of the other code in this PR.